docs/upgrade.md

```diff
@@ -3,5 +3,263 @@ title: Upgrade guide
+const myApiKey = await getSecret('my-api-key', { transform: 'json' });
+
+export const handler = async () => {
+    // ...
+};
+```
+
+With this change, you can also apply tree-shaking to your function bundle to reduce its size. As part of this release we have made changes to the package and its exports to better support this feature, and we remain committed to improving this in the future based on your feedback.
+
+While we recommend using ES Modules, we understand that this change might not be possible for everyone. If you're unable to use ES Modules, you can continue to use the `require` syntax to import the package. Powertools for AWS Lambda (TypeScript) will continue to support this syntax by shipping CommonJS modules alongside ES Modules.
+
+In some cases, even when opting for ES Modules, you might still need to use the `require` syntax to import a package. For example, if you're using a package that doesn't support ES Modules, or if one of your transitive dependencies is using the `require` syntax like it's the case for `@aws-lambda-powertools/tracer` which relies on the AWS X-Ray SDK for Node.js. In these cases, you can still use ES Modules for the rest of your codebase and set a special build flag to tell your bundler to inject a banner at the top of the file to use the `require` syntax for the specific package.
+
+=== "With AWS CDK"
+
+    ```typescript hl_lines="15 20-21"
+    import { Stack, type StackProps } from 'aws-cdk-lib';
+    import { Construct } from 'constructs';
+    import { NodejsFunction, OutputFormat } from 'aws-cdk-lib/aws-lambda-nodejs';
+    import { Runtime } from 'aws-cdk-lib/aws-lambda';
+
+    export class MyStack extends Stack {
+      public constructor(scope: Construct, id: string, props?: StackProps) {
+        super(scope, id, props);
+
+        const handler = new NodejsFunction(this, 'helloWorldFunction', {
+          runtime: Runtime.NODEJS_20_X,
+          handler: 'handler',
+          entry: 'src/index.ts',
+          bundling: {
+            format: OutputFormat.ESM,
+            minify: true,
+            esbuildArgs: {
+              "--tree-shaking": "true",
+            },
+            banner: 
+              "import { createRequire } from 'module';const require = createRequire(import.meta.url);",
+          },
+        });
+      }
+    }
+    ```
+
+=== "With AWS SAM"
+
+    ```yaml hl_lines="14 17-18"
+    Transform: AWS::Serverless-2016-10-31
+    Resources:
+      HelloWorldFunction:
+        Type: AWS::Serverless::Function
+        Properties:
+          Runtime: nodejs20.x
+          Handler: src/index.handler
+        Metadata:
+          BuildMethod: esbuild
+          BuildProperties:
+            Minify: true
+            Target: 'ES2020'
+            Sourcemap: true
+            Format: esm
+            EntryPoints:
+              - src/index.ts
+            Banner:
+              js: "import { createRequire } from 'module';const require = createRequire(import.meta.url);"
+ 
+    ```
```

Comment:
```suggestion
              js: "import { createRequire } from 'module';const require = createRequire(import.meta.url);"  # (1)!
 
    ```

    1. `esbuild` will include this arbitrary code at the top of your bundle to maximize CommonJS compatibility _(`require` keyword)_.
```